File: answers/leetcode/Intersection of Two Linked Lists/Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """

        lena = 0
        currenta = headA
        while currenta:
            lena += 1
            currenta = currenta.next
        lenb = 0
        currentb = headB
        while currentb:
            lenb += 1
            currentb = currentb.next

        currenta = headA
        currentb = headB
        for i in range(min(lena, lenb), max(lena, lenb)):
            if lena > lenb:
                currenta = currenta.next
            else:
                currentb = currentb.next
        while currenta and currentb and currenta != currentb:
            currenta = currenta.next
            currentb = currentb.next
        return currenta

File: answers/leetcode/Intersection of Two Linked Lists/Intersection of Two Linked Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int diff = 0;
        ListNode currentA = headA;
        while (currentA != null) {
            diff++;
            currentA = currentA.next;
        }
        ListNode currentB = headB;
        while (currentB != null) {
            diff--;
            currentB = currentB.next;
        }
        currentA = headA;
        currentB = headB;
        while (diff > 0) {
            currentA = currentA.next;
            diff--;
        }
        while (diff < 0) {
            currentB = currentB.next;
            diff++;
        }
        while (currentA != currentB) {
            currentA = currentA.next;
            currentB = currentB.next;
        }
        return currentA;
    }
}